File: apps/backend-python/src/services/note_service.py
import json
from datetime import datetime
from nanoid import generate
from sqlalchemy.orm import Session
from src.db.models import Note
from src.utils.errors import NotFoundError


def _serialize(note: Note) -> dict:
    return {
        "id": note.id,
        "userId": note.user_id,
        "title": note.title,
        "content": note.content,
        "plainText": note.plain_text,
        "visibility": note.visibility,
        "wordCount": note.word_count,
        "charCount": note.char_count,
        "tags": json.loads(note.tags) if isinstance(note.tags, str) and note.tags else [],
        "folderId": note.folder_id,
        "archived": note.archived,
        "pinned": note.pinned,
        "shareId": note.share_id,
        "createdAt": note.created_at.isoformat() if note.created_at else None,
        "updatedAt": note.updated_at.isoformat() if note.updated_at else None,
    }


def _strip_html(html: str) -> str:
    import re
    return re.sub(r"<[^>]+>", "", html)


class NoteService:
    def create_note(self, db: Session, user_id: str, data: dict) -> dict:
        now = datetime.utcnow()
        content = data.get("content", "")
        plain = _strip_html(content)
        note = Note(
            id=generate(), user_id=user_id,
            title=data["title"], content=content, plain_text=plain,
            visibility=data.get("visibility", "private"),
            word_count=len(plain.split()),
            char_count=len(plain),
            tags=json.dumps(data.get("tags", [])),
            folder_id=data.get("folderId"),
            created_at=now, updated_at=now,
        )
        db.add(note)
        db.commit()
        db.refresh(note)
        return _serialize(note)

    def get_notes(self, db: Session, user_id: str, filters: dict = None) -> list:
        q = db.query(Note).filter(Note.user_id == user_id)
        if filters:
            if "archived" in filters:
                q = q.filter(Note.archived == filters["archived"])
            if "pinned" in filters:
                q = q.filter(Note.pinned == filters["pinned"])
            if "folderId" in filters:
                q = q.filter(Note.folder_id == filters["folderId"])
            if "search" in filters:
                term = f"%{filters['search']}%"
                q = q.filter(Note.title.ilike(term) | Note.plain_text.ilike(term))
        notes = q.order_by(Note.updated_at.desc()).all()
        return [_serialize(n) for n in notes]

    def get_note_by_id(self, db: Session, note_id: str, user_id: str) -> dict:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        return _serialize(note)

    def get_note_by_share_id(self, db: Session, share_id: str) -> dict:
        note = db.query(Note).filter(Note.share_id == share_id).first()
        if not note:
            raise NotFoundError("Note not found")
        return _serialize(note)

    def update_note(self, db: Session, note_id: str, user_id: str, data: dict) -> dict:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        if "title" in data:
            note.title = data["title"]
        if "content" in data:
            note.content = data["content"]
            plain = _strip_html(data["content"])
            note.plain_text = plain
            note.word_count = len(plain.split())
            note.char_count = len(plain)
        if "visibility" in data:
            note.visibility = data["visibility"]
        if "tags" in data:
            note.tags = json.dumps(data["tags"])
        if "folderId" in data:
            note.folder_id = data["folderId"]
        note.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(note)
        return _serialize(note)

    def delete_note(self, db: Session, note_id: str, user_id: str) -> None:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        db.delete(note)
        db.commit()

    def archive_note(self, db: Session, note_id: str, user_id: str, archived: bool) -> dict:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        note.archived = archived
        note.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(note)
        return _serialize(note)

    def pin_note(self, db: Session, note_id: str, user_id: str, pinned: bool) -> dict:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        note.pinned = pinned
        note.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(note)
        return _serialize(note)

    def share_note(self, db: Session, note_id: str, user_id: str) -> dict:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        if not note.share_id:
            note.share_id = generate()
            note.updated_at = datetime.utcnow()
            db.commit()
            db.refresh(note)
        return {"shareId": note.share_id}

    def unshare_note(self, db: Session, note_id: str, user_id: str) -> None:
        note = db.query(Note).filter(Note.id == note_id, Note.user_id == user_id).first()
        if not note:
            raise NotFoundError("Note not found")
        note.share_id = None
        note.updated_at = datetime.utcnow()
        db.commit()


note_service = NoteService()

File: apps/backend/src/services/note.service.ts
import { nanoid } from 'nanoid';
import { eq, and, desc, like, or } from 'drizzle-orm';
import { db } from '../config/database';
import { notes } from '../db/schema';
import { NotFoundError } from '../utils/errors';

function htmlToPlainText(html: string): string {
  return html
    .replace(/<[^>]*>/g, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/\s+/g, ' ')
    .trim();
}

function countWords(text: string): number {
  if (!text.trim()) return 0;
  return text.trim().split(/\s+/).length;
}

export const noteService = {
  async createNote(userId: string, data: {
    title: string;
    content?: string;
    visibility?: 'private' | 'public' | 'shared';
    tags?: string[];
    folderId?: string | null;
  }) {
    const noteId = nanoid();
    const now = new Date();
    const content = data.content || '';
    const plainText = htmlToPlainText(content);

    await db.insert(notes).values({
      id: noteId,
      userId,
      title: data.title,
      content,
      plainText,
      visibility: data.visibility || 'private',
      wordCount: countWords(plainText),
      charCount: plainText.length,
      tags: data.tags ? JSON.stringify(data.tags) : null,
      folderId: data.folderId || null,
      archived: false,
      pinned: false,
      shareId: null,
      createdAt: now,
      updatedAt: now,
    });

    return this.getNoteById(noteId, userId);
  },

  async getNotes(userId: string, filters?: {
    archived?: boolean;
    pinned?: boolean;
    folderId?: string;
    search?: string;
    page?: number;
    limit?: number;
  }) {
    let query = db.select().from(notes).where(eq(notes.userId, userId));

    const conditions = [eq(notes.userId, userId)];

    if (filters?.archived !== undefined) {
      conditions.push(eq(notes.archived, filters.archived));
    }

    if (filters?.pinned !== undefined) {
      conditions.push(eq(notes.pinned, filters.pinned));
    }

    if (filters?.folderId !== undefined) {
      conditions.push(eq(notes.folderId, filters.folderId));
    }

    if (filters?.search) {
      const searchTerm = `%${filters.search}%`;
      conditions.push(
        or(
          like(notes.title, searchTerm),
          like(notes.plainText, searchTerm)
        )!
      );
    }

    const result = await db.select()
      .from(notes)
      .where(and(...conditions))
      .orderBy(desc(notes.pinned), desc(notes.updatedAt));

    
    return result.map(note => ({
      ...note,
      tags: note.tags ? JSON.parse(note.tags) : [],
    }));
  },

  async getNoteById(id: string, userId: string) {
    const [note] = await db.select()
      .from(notes)
      .where(and(eq(notes.id, id), eq(notes.userId, userId)))
      .limit(1);

    if (!note) {
      throw new NotFoundError('Note not found');
    }

    return {
      ...note,
      tags: note.tags ? JSON.parse(note.tags) : [],
    };
  },

  async getNoteByShareId(shareId: string) {
    const [note] = await db.select()
      .from(notes)
      .where(eq(notes.shareId, shareId))
      .limit(1);

    if (!note) {
      throw new NotFoundError('Shared note not found');
    }

    return {
      ...note,
      tags: note.tags ? JSON.parse(note.tags) : [],
    };
  },

  async updateNote(id: string, userId: string, data: {
    title?: string;
    content?: string;
    visibility?: 'private' | 'public' | 'shared';
    tags?: string[];
    folderId?: string | null;
  }) {
    await this.getNoteById(id, userId);

    const updateData: any = {
      updatedAt: new Date(),
    };

    if (data.title !== undefined) {
      updateData.title = data.title;
    }

    if (data.content !== undefined) {
      updateData.content = data.content;
      updateData.plainText = htmlToPlainText(data.content);
      updateData.wordCount = countWords(updateData.plainText);
      updateData.charCount = updateData.plainText.length;
    }

    if (data.visibility !== undefined) {
      updateData.visibility = data.visibility;
    }

    if (data.tags !== undefined) {
      updateData.tags = JSON.stringify(data.tags);
    }

    if (data.folderId !== undefined) {
      updateData.folderId = data.folderId;
    }

    await db.update(notes)
      .set(updateData)
      .where(eq(notes.id, id));

    return this.getNoteById(id, userId);
  },

  async deleteNote(id: string, userId: string) {
    await this.getNoteById(id, userId);
    await db.delete(notes).where(eq(notes.id, id));
  },

  async archiveNote(id: string, userId: string, archived: boolean) {
    await this.getNoteById(id, userId);

    await db.update(notes)
      .set({ archived, updatedAt: new Date() })
      .where(eq(notes.id, id));

    return this.getNoteById(id, userId);
  },

  async pinNote(id: string, userId: string, pinned: boolean) {
    await this.getNoteById(id, userId);

    await db.update(notes)
      .set({ pinned, updatedAt: new Date() })
      .where(eq(notes.id, id));

    return this.getNoteById(id, userId);
  },

  async shareNote(id: string, userId: string) {
    await this.getNoteById(id, userId);

    const shareId = nanoid();

    await db.update(notes)
      .set({ shareId, visibility: 'public', updatedAt: new Date() })
      .where(eq(notes.id, id));

    return { shareId };
  },

  async unshareNote(id: string, userId: string) {
    await this.getNoteById(id, userId);

    await db.update(notes)
      .set({ shareId: null, visibility: 'private', updatedAt: new Date() })
      .where(eq(notes.id, id));
  },
};
